miqa/core/rest/session.py

```diff
@@ -70,52 +72,59 @@ def import_(self, request, **kwargs):
-                Response({'error': f'Invalid CSV file: {str(e)}'})
+                raise ValidationError({'error': f'Invalid CSV file: {str(e)}'})
 
-        print(json_content)
-        data_root = json_content['data_root']
+        data_root = Path(json_content['data_root'])
 
-        for site in json_content['sites']:
-            # TODO a different user will result in a duplicate site
-            Site.objects.get_or_create(name=site['id'], session=session, creator=request.user)
+        sites = {
+            site['name']: Site.objects.get_or_create(
+                name=site['name'], defaults={'creator': request.user}
+            )[0]
+            for site in json_content['sites']
+        }
 
-        for experiment in json_content['experiments']:
-            # TODO a different user or note will result in a duplicate experiment
-            Experiment.objects.get_or_create(
-                name=experiment['id'],
-                note=experiment['note'],
-                session=session,
-            )
-        for scan in json_content['scans']:
-            experiment = Experiment.objects.get(name=scan['experiment_id'], session=session)
-            site = Site.objects.get(name=scan['site_id'], session=session, creator=request.user)
-            decision = ScanDecision.from_rating(scan['decision'])
-            scan_model, _created = Scan.objects.get_or_create(
-                scan_id=scan['id'],
-                scan_type=scan['type'],
-                decision=decision,
-                note=scan['note'],
+        Experiment.objects.filter(session=session).delete()  # cascades to scans -> images
+
+        experiments = {
+            e['id']: Experiment(name=e['id'], note=e['note'], session=session)
+            for e in json_content['experiments']
+        }
+        Experiment.objects.bulk_create(experiments.values())
+
+        scans = []
+        images = []
+        for scan_json in json_content['scans']:
+            experiment = experiments[scan_json['experiment_id']]
+            site = sites[scan_json['site_id']]
+            scan = Scan(
+                scan_id=scan_json['id'],
+                scan_type=scan_json['type'],
+                decision=ScanDecision.from_rating(scan_json['decision']),
+                note=scan_json['note'],
                 experiment=experiment,
                 site=site,
-                # TODO session?
             )
-            if 'images' in scan:
+            scans.append(scan)
+
+            if 'images' in scan_json:
                 # TODO implement this
-                raise Exception('use imagePattern for now')
-            # TODO change this to image_pattern in the schema
-            elif 'imagePattern' in scan:
-                image_pattern = re.compile(scan['imagePattern'])
-                image_dir = os.path.join(data_root, scan['path'])
+                raise Exception('use image_pattern for now')
+            elif 'image_pattern' in scan_json:
```

Comment:
We need to make sure this change in column header gets communicated to whoever is writing the CSVs.

Reply:
This key only exists in the JSON, right? I don't think it appears in any of the CSVs